src/Bicep.Core/PrettyPrint/Documents/NestDocument.cs

```diff
@@ -1,45 +1,60 @@
 ﻿// Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
+using System.Collections.Immutable;
+using System.Linq;
+using System.Runtime.CompilerServices;
 using System.Text;
-using Bicep.Core.PrettyPrint.Documents;
 
-namespace Bicep.Core.PrettyPrint.DocumentCombinators
+namespace Bicep.Core.PrettyPrint.Documents
 {
     public class NestDocument : ILinkedDocument
     {
-        public NestDocument(int level, ILinkedDocument successor)
+        private readonly int level;
+
+        private readonly ImmutableArray<ILinkedDocument> successors;
+
+        public NestDocument(int level)
+            : this(level, ImmutableArray<ILinkedDocument>.Empty)
         {
-            this.Level = level;
-            this.Successor = successor;
         }
 
-        public int Level { get; }
-
-        public ILinkedDocument Successor { get; }
+        public NestDocument(int level, ImmutableArray<ILinkedDocument> successors)
+        {
+            this.level = level;
+            this.successors = successors;
+        }
 
         public ILinkedDocument Concat(ILinkedDocument other)
         {
-            return new NestDocument(this.Level, this.Successor.Concat(other));
+            return new NestDocument(this.level, this.successors.Add(other));
         }
 
-        public ILinkedDocument Nest(int level)
+        public ILinkedDocument Nest()
         {
-            return new NestDocument(level + this.Level, this.Successor.Nest(level));
+            RuntimeHelpers.EnsureSufficientExecutionStack();

```

Comment:
>RuntimeHelpers.EnsureSufficientExecutionStack() [](start = 12, length = 47)

Is the `DocumentBuildVisitor` ultimately driving the recursion here? If yes, would it make sense to have one call to `RuntimeHelpers.EnsureSufficientExecutionStack()` there?

Reply:
Otherwise, we'd have to stick this everywhere :(

---
In reply to: [585109461](https://github.com/Azure/bicep/pull/1668#discussion_r585109461) [](ancestors = 585109461)

Reply:
`DocumentBuildVisitor` is driving the recursion, but if I understand it correctly putting it in the parent method call won't help, and it has to be put inside a recursion method body.